builtins: fix crdb_internal.encode_key for user defined types

otan opened the pull request:
Resolves #63646

Release note (bug fix): Fix crdb_internal.encode_key for user defined
types. This would previously error. This will also fix SHOW RANGE FOR
ROW for tables with user-defined types as their PRIMARY KEY.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63716)
<!-- Reviewable:end -->

otan commented:
tagging you on this @ajwerner because you have the gross TODO which i made even more gross :P

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63716)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/sql/sem/builtins/builtins.go, line 3963 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63716#-MYKluuy3IcbXQanZC7e:-MYKluuy3IcbXQanZC7f:b-ao76jx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7e9fdbe379511b0ca2a9942493c770e4de0c9ed/pkg/sql/sem/builtins/builtins.go#L3963)):*
> ```Go
> 				// TODO(ajwerner): This is awful, we should be able to resolve this
> 				// thing using the usual tools rather than going through the DB.
> 				tableDesc, err := catalogkv.MustGetTableDescByID(ctx.Context, ctx.Txn, ctx.Codec, descpb.ID(tableID))
> ```

Rather than pile yet another sin into this builtin, let's fix the original sin here. Nothing should be using catalogkv here. We should be using a descriptor collection. As it stands in this change, this function is going to cost 2 round-trips per invocation. That's very bad. FWIW I'm not super keen on all of this business logic living here anyway. My relatively strong preference would be to have a corresponding method hanging off of the planner and invoked through an interface. That way you don't need to plumb more dependencies up to the `builtins` package. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
I'm going through and getting rid of all of the catalogkv calls. I have long wanted to make catalogkv internal to catalog. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63716#-:-MYKorThDwBHaA57s9Am:bhiw8se)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)



<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63716)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[pkg/sql/sem/builtins/builtins.go, line 3963 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63716#-MYKluuy3IcbXQanZC7e:-MYMeN1DB19zjFkxNEbC:b-7rl0jq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7e9fdbe379511b0ca2a9942493c770e4de0c9ed/pkg/sql/sem/builtins/builtins.go#L3963)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Rather than pile yet another sin into this builtin, let's fix the original sin here. Nothing should be using catalogkv here. We should be using a descriptor collection. As it stands in this change, this function is going to cost 2 round-trips per invocation. That's very bad. FWIW I'm not super keen on all of this business logic living here anyway. My relatively strong preference would be to have a corresponding method hanging off of the planner and invoked through an interface. That way you don't need to plumb more dependencies up to the `builtins` package. 
</blockquote></details>

done!


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks

 I’d prefer if even more of the business logic were in the sql package but appreciate the bug fix. I coming to think of the builtins package like a router to deal with decode arguments and then pass them to an implementation. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63716#-:-MYMkI9l4pFsMbGpN4AE:bvu8ol0)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[pkg/sql/sem/builtins/builtins.go, line 3963 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63716#-MYKluuy3IcbXQanZC7e:-MYMjxc--EgBi3rSAX_c:bz5j4l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7e9fdbe379511b0ca2a9942493c770e4de0c9ed/pkg/sql/sem/builtins/builtins.go#L3963)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

done!
</blockquote></details>

Heh I was imagining that you’d put a method on the planner called something like EncodeKey but this is better than where we were before. 


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
i'll go with this for now ;)

bors r=ajwerner

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63716#-:-MYMpitvDFNcvJrdcCF9:b910ea3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[pkg/sql/sem/builtins/builtins.go, line 3963 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63716#-MYKluuy3IcbXQanZC7e:-MYMpWqj3HKtmm9ceQLc:b6nsu0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7e9fdbe379511b0ca2a9942493c770e4de0c9ed/pkg/sql/sem/builtins/builtins.go#L3963)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Heh I was imagining that you’d put a method on the planner called something like EncodeKey but this is better than where we were before. 
</blockquote></details>

i wish the builtins package took in something more than evalctx, so we can make builtins richer from stuff that shouldn't be there, but here we are!


<!-- Sent from Reviewable.io -->

otan commented:
bors r-

otan commented:
bors r=ajwerner

otan committed:
builtins: fix crdb_internal.encode_key for user defined types

Release note (bug fix): Fix crdb_internal.encode_key for user defined
types. This would previously error. This will also fix SHOW RANGE FOR
ROW for tables with user-defined types as their PRIMARY KEY.

otan commented:
bors r=ajwerner
